refactor (#322)
KxSystems/docs: docs/basics/ipc.md

[File system](files.md)

## Listen for connections

A kdb+ process can define a [listening port](listening-port.md) at start-up or at runtime. 

> kdb+ can receive messages over TCP, UDS (unix domain sockets), named pipes or a range of 3rd party middlewares (e.g. Kafka, Solace, etc)

Reviewer: ```suggestion
kdb+ can receive messages over TCP, UDS (unix domain sockets), named pipes or a range of third party middlewares (for example, Kafka, Solace, and so on).
```

## Connecting

> A kdb+ process can connect to another using [`hopen`](../ref/hopen.md) e.g: to start a kdb+ process listening on port 5000

Reviewer: ```suggestion
A kdb+ process can connect to another using [`hopen`](../ref/hopen.md). For example, to start a kdb+ process listening on port 5000.
```

```bash
$ q -p 5001
```
another kdb+ process can connect to this process with [`hopen`](../ref/hopen.md):

```q
q)h:hopen `::5001
q)h                 /h is the socket (an OS file descriptor)
3i
```

Sync messages can also be sent without a pre-existing connection using [one-shot](#one-shot-message).

> The max number of connections is defined by the system limit for protocol (operating system configurable). Prior to 4.1t 2023.09.15, the limit was hardcoded to 1022.

Reviewer: ```suggestion
The maximum number of connections is defined by the system limit for protocol (operating system configurable). Prior to 4.1t 2023.09.15, the limit was hardcoded to 1022.
```

Send messages from the client to the server using the [connection handle](handles.md) returned from `hopen`.

There are three message types: sync, async, and response.

### Sync request (get)

Performs the following:

1. sends any pending outgoing (async) messages on `h`
1. sends the sync request message
1. processes any pending incoming messages on `h` until a response (or error) message is received

```q
q)h"2+2" / this is sent to the remote process for calculation
4
```

> The basic method used to execute a query via IPC is sending the query as a string as in the above example. 

Reviewer: ```suggestion
The recommended method to execute a query through IPC is sending the query as a string as shown in the above example. 
```

Author: It wouldn't be the recommended method. Perhaps something like "The method shown above is sending the query as a string."

Author: changed wording to "The method shown above is sending the query as a string." in latest commit below

> A function can also be executed on the server by passing a [parse tree](parsetrees.md) to the handle: a list with the function as first item, followed by its arguments.

Reviewer: ```suggestion
You can also execute a function on the server by passing a [parse tree](parsetrees.md) to the handle: a list with the function as first item, followed by its arguments.
```

> To execute a function defined on the *client side*, simply pass the function name so it will be resolved before sending. 

Reviewer: ```suggestion
To execute a function defined on the *client side*, pass the function name so it is resolved before sending. 
```

A useful shorthand for a one-shot get is:

```q
q)`::5001 "1+1"
2
```

Since V4.0 2020.03.09, a one-shot query can be run with a timeout (in milliseconds), as in the second example below:

```q
q)`::4567"2+2"
4
q)`::[(`::4567;100);"1+1"]
2
```

#### Interrupting requests

It is possible to interrupt a long-running sync query with `kill -s INT *PID*`. As with the previous example, any subsequent attempt to communicate across this handle will fail.

```q
q)h"system\"sleep 30\""
'rcv handle: 4. OS reports: Interrupted system call
  [0]  h"system\"sleep 30\""
       ^
q)
q)h"a"
'Cannot write to handle 4. OS reports: Bad file descriptor
  [0]  h"a"
       ^
```

### Response message (get response)

Sent automatically by the listening process on completing a sync (get) request.

### Async message (set)

Serializes and puts a message on the output queue for handle `h`, and does not block client nor wait for any response message. A negative handle signifies async.

```q
q)neg[h]"a:10" / on the remote instance, sets the variable a to 10
```

As per [sync messages](#sync-request-get), you can replace the string repesenting the code to execute as a list that forms a parse tree.

> Since the process is not waiting for a response, async querying is critical in situations where waiting for an unresponsive subscriber is unacceptable, e.g. in a tickerplant.

Reviewer: ```suggestion
Since the process is not waiting for a response, async querying is critical in situations where waiting for an unresponsive subscriber is unacceptable, for example, in a tickerplant.
```

You may consider increasing the size of TCP send/receive buffers on your system to reduce the amount of blocking whilst trying to write into a socket.

#### Flushing

Messages can be queued for sending to a remote process through using async messaging. 

> kdb+ will queue the serialized message in user space, later writing it to the socket as the remote end drains the message queue. 

Reviewer: ```suggestion
kdb+ queues the serialized message in user space, later writing it to the socket as the remote end drains the message queue. 
```

> One can see how many messages are queued on a handle and their sizes as a dictionary through the command variable [`.z.W`](../ref/dotz.md#zw-handles "handles").

Reviewer: ```suggestion
You can see how many messages are queued on a handle and their sizes as a dictionary using the command variable [`.z.W`](../ref/dotz.md#zw-handles "handles").
```

Sometimes it is useful to send a large number of aysnc messages, but then to block until they have all been sent. 
This can be achieved through using async flush – invoked as `neg[h][]` or `neg[h](::)`. 

> If you need confirmation that the remote end has received and processed the async messages, chase them with a sync request, 
> e.g. `h""` – the remote end will process the messages on a socket in the order that they are sent.

Reviewer: ```suggestion
If you need confirmation that the remote end has received and processed the async messages, use a sync request. For example, `h""` – the remote end processes the messages on a socket in the order that they are sent.
```

> !!! note "flushing can also be achieved by sending a synchronous message on the same handle: this will confirm execution as all messages are processed in the order they are sent"

Reviewer: ```suggestion
!!! note "Flushing can also be achieved by sending a synchronous message on the same handle. This confirms execution as all messages are processed in the order they are sent."
```

#### Broadcast

> Much of the overhead of sending a message via IPC is in serializing the data before sending. 
> It is possible to ‘async broadcast’ the same message to multiple handles using the internal [-25!](internal.md#-25x-async-broadcast) function. 
> This will serialize the message once and send to all handles to reduce CPU and memory load.

Reviewer: ```suggestion
Much of the overhead of sending a message using IPC is in serializing the data before sending. 
It is possible to ‘async broadcast’ the same message to multiple handles using the internal [-25!](internal.md#-25x-async-broadcast) function. This serializes the message once and send to all handles to reduce CPU and memory load.
```

:fontawesome-solid-graduation-cap: 
[Using `.z`](../kb/using-dotz.md) for more resources, including contributed code for tracing and monitoring

### Async blocking

To block until any async message is received on handle `h`

```q
r:h[] / store message in r
```

## Authentication / Authorization

> Basic access control and authentication is supported through using the [-u](cmdline.md#-u-usr-pwd-local)/[-U](cmdline.md#-u-usr-pwd) command-line option to specify a file of users and passwords.

Reviewer: ```suggestion
Basic access control and authentication is supported by using the [-u](cmdline.md#-u-usr-pwd-local)/[-U](cmdline.md#-u-usr-pwd) command-line option to specify a file of users and passwords.
```

> In order to provide further customizations of the authentication process, [.z.pw](../ref/dotz.md#zpw-validate-user) callback is called immediately after successful –u/-U authentication (if specified at startup – otherwise .z.pw is the first authentication check done by a kdb+ process). The ability to set .z.pw to user defined function, allows allows integration with enterprise standards such as LDAP, Kerberos, OpenID Connect,etc

Reviewer: ```suggestion
To further customize the authentication process, [.z.pw](../ref/dotz.md#zpw-validate-user) callback is called immediately after successful –u/-U authentication, if specified at start up. The ability to set .z.pw to user defined function, allows allows integration with enterprise standards such as LDAP, Kerberos, OpenID Connect, etc.
```

Author: Suggestion changes meaning. .z.pw is called on initial connection, or if no -u/-U defined its called for each new connection. Will submit new change

> Finer grained authorization can be implemented by tracking user info with active handles, and customizing sync/async callbacks for user-level permissioning e.g. server with protected funcs for sync calls

Reviewer: ```suggestion
Finer grained authorization can be implemented by tracking user information with active handles and customizing sync/async callbacks for user-level permissioning, for example, server with protected functions for sync calls.
```

```q
q)\p 5000
q)allowedFns:(`func1;`func2;`func3;+;-) / list of allowed function/ops to call
q)checkFn:{if[not x in allowedFns;'(-3!x)," not allowed"];}
q)validatePT:{if[0h=t:type x;if[(not 0h=type first x)&1=count first x;checkFn first x;];.z.s each x where 0h=type each x;];}
q).z.pg:{if[10h=type x;x:parse x;];validatePT x;eval x}
```

client trying to access protected functions:
```q
q)h:hopen 5000
q)h"1+1"
2
q)h"1*1"
'* not allowed
  [0]  h"1*1"
       ^
```

!!! note

>     ticker plants, and other high-volume message sources such as feed handlers, will most likely be inserting data via .z.ps. To cater for such high volumes, the handles of those processes should be used to avoid the overhead of these validation checks. That is, feeds and tickerplants could be viewed as trusted processes

Reviewer: ```suggestion
    Ticker plants and other high-volume message sources, such as feed handlers, generally insert data using `.z.ps`. To manage such high volumes, the handles of those processes should be used to avoid the overhead of these validation checks. That is, feeds and tickerplants could be viewed as trusted processes.
```

:fontawesome-regular-map:
[Permissions with kdb+](../wp/permissions/index.md "White paper")
<br>
:fontawesome-regular-map:
[Using .z](../kb/using-dotz.md)

## Tracking connections

A list of current connections can be viewed using [.z.H](../ref/dotz.md#zh-active-sockets). A more detailed list is achieved via [-38!](internal.md#-38x-socket-table).

> Further tracking of connections on a server (tracking client connections) can be accomplished using customized implementations of [.z.po](../ref/dotz.md#zpo-open) and [.z.pc](../ref/dotz.md#zpc-close)

Reviewer: ```suggestion
Further tracking of connections on a server (tracking client connections) can be accomplished using customized implementations of [.z.po](../ref/dotz.md#zpo-open) and [.z.pc](../ref/dotz.md#zpc-close).
```

## Protocol

The protocol is extremely simple, as is the message format. 

> One can see what a TCP/IP message looks like by using [`-8!object`](internal.md#-8x-to-bytes), which generates the byte vector for the [serialization](../kb/serialization.md) of the object.

Reviewer: ```suggestion
You can see what a TCP/IP message looks like by using [`-8!object`](internal.md#-8x-to-bytes), which generates the byte vector for the [serialization](../kb/serialization.md) of the object.
```
